### ISSUE TITLE
Vec.hashCode gives a lot of conflicts

### ISSUE THREAD

**Issue by filiph** [munificent/piecemeal#4]
```dart
int get hashCode => (x ^ y).hashCode;
```

https://github.com/munificent/piecemeal/blob/master/lib/src/vec.dart#L259

This basically guarantees conflicts. For example, here's a list of vectors in a 100x100 field that evaluate to `Vec.hashCode == 1`:

```
(0, 1), (1, 0), (2, 3), (3, 2), (4, 5), (5, 4), (6, 7), (7, 6), (8, 9), (9, 8), (10, 11), (11, 10), (12, 13), (13, 12), (14, 15), (15, 14), (16, 17), (17, 16), (18, 19), (19, 18), (20, 21), (21, 20), (22, 23), (23, 22), (24, 25), (25, 24), (26, 27), (27, 26), (28, 29), (29, 28), (30, 31), (31, 30), (32, 33), (33, 32), (34, 35), (35, 34), (36, 37), (37, 36), (38, 39), (39, 38), (40, 41), (41, 40), (42, 43), (43, 42), (44, 45), (45, 44), (46, 47), (47, 46), (48, 49), (49, 48), (50, 51), (51, 50), (52, 53), (53, 52), (54, 55), (55, 54), (56, 57), (57, 56), (58, 59), (59, 58), (60, 61), (61, 60), (62, 63), (63, 62), (64, 65), (65, 64), (66, 67), (67, 66), (68, 69), (69, 68), (70, 71), (71, 70), (72, 73), (73, 72), (74, 75), (75, 74), (76, 77), (77, 76), (78, 79), (79, 78), (80, 81), (81, 80), (82, 83), (83, 82), (84, 85), (85, 84), (86, 87), (87, 86), (88, 89), (89, 88), (90, 91), (91, 90), (92, 93), (93, 92), (94, 95), (95, 94), (96, 97), (97, 96), (98, 99), (99, 98)
```

A better solution would be — &#42;checks stackoverflow&#42; — Szudzik's function!

Something like:
```dart
  int get hashCode {
    // See: https://stackoverflow.com/a/13871379/1416886
    var A = x >= 0 ? 2 * x : -2 * x - 1;
    var B = y >= 0 ? 2 * y : -2 * y - 1;
    var C = (A >= B ? A * A + A + B : A + B * B) ~/ 2;
    return x < 0 && y < 0 || x >= 0 && y >= 0 ? C : -C - 1;
  }
```

I can make the PR.

**Comment by munificent** [comment]
That's pretty gnarly, but I think this would work well:

```dart
int hashPoint(int x, int y, [int seed]) {
  seed ??= 0;

  // From: https://stackoverflow.com/a/12996028/9457
  hashInt(int n) {
    n = (((n >> 16) ^ n) * 0x45d9f3b) & 0xffffffff;
    n = (((n >> 16) ^ n) * 0x45d9f3b) & 0xffffffff;
    n = (n >> 16) ^ n;
    return n;
  }

  return hashInt(hashInt(hashInt(seed) + x) + y) & 0xffffffff;
}
```

It's based on https://stackoverflow.com/a/12996028/9457. When I get a chance, I'll roll this in.

**Comment by filiph** [comment]
I tried to test this and this still makes a lot of duplicates, while the Szudzik's function does not. Here's how I test it (a field of 2000x2000):

```dart
main() {
  int currentHash(int x, int y) => (x ^ y).hashCode;

  int szudzikHash(int x, int y) {
    var A = x >= 0 ? 2 * x : -2 * x - 1;
    var B = y >= 0 ? 2 * y : -2 * y - 1;
    var C = (A >= B ? A * A + A + B : A + B * B) ~/ 2;
    return x < 0 && y < 0 || x >= 0 && y >= 0 ? C : -C - 1;
  }

  int hashPoint(int x, int y, [int seed]) {
    seed ??= 0;

    // From: https://stackoverflow.com/a/12996028/9457
    hashInt(int n) {
      n = (((n >> 16) ^ n) * 0x45d9f3b) & 0xffffffff;
      n = (((n >> 16) ^ n) * 0x45d9f3b) & 0xffffffff;
      n = (n >> 16) ^ n;
      return n;
    }

    return hashInt(hashInt(hashInt(seed) + x) + y) & 0xffffffff;
  }

  final hashes = Set<int>();

  int range = 1000;

  for (int i = -range; i < range; i++) {
    for (int j = -range; j < range; j++) {
      final hash = hashPoint(i, j);
      if (hashes.contains(hash)) {
        print('duplicate: ($i, $j) - $hash');
      }
      hashes.add(hash);
    }
  }
}
```

I also tested performance (removing the print statement) and your approach seems to be faster by about 20%.

**Comment by munificent** [comment]
You've effectively nerd sniped me. Here's one I like better:

```dart
int cantorHash(int x, int y) {
  x = x >= 0 ? 2 * x : -2 * x - 1;
  y = y >= 0 ? 2 * y : -2 * y - 1;
  return (x * x + 3 * x + 2 * x * y + y + y * y) ~/ 2;
}
```

This is based on the [Cantor pairing function](https://en.wikipedia.org/wiki/Pairing_function) the paper you mentioned references. The first two lines do the same trick as Szudzik does to map negative coordinates into positive coordinate space (and to spread out the positive ones to make room for them). The last line is the Cantor pairing function.

Like Szudzik's solution, this has no collisions, at least for relatively small integers. Eventually you'll get collisions when multiplications overflow.

With the benchmark you wrote, this is about 2x faster, but it's not entirely clear to me what's being measured. Almost all of the time goes into adding the hashes to the set. In theory, the Cantor hash should be faster since it's doing fewer branches and less math overall.

**Comment by munificent** [comment]
Fixed in 71aa6567b75aad358328ee8eab0f062c1ea3ed2d.
